source/extensions/http/cache/file_system_http_cache/cache_eviction_thread.cc

```diff
@@ -0,0 +1,188 @@
+#include "source/extensions/http/cache/file_system_http_cache/cache_eviction_thread.h"
+
+#include <limits>
+
+#include "envoy/thread/thread.h"
+
+#include "source/common/api/os_sys_calls_impl.h"
+#include "source/common/filesystem/directory.h"
+#include "source/extensions/http/cache/file_system_http_cache/file_system_http_cache.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Cache {
+namespace FileSystemHttpCache {
+
+namespace {
+bool isCacheFile(const Filesystem::DirectoryEntry& entry) {
+  return entry.type_ == Filesystem::FileType::Regular && absl::StartsWith(entry.name_, "cache-");
+}
+} // namespace
+
+CacheEvictionThread::CacheEvictionThread(Thread::ThreadFactory& thread_factory)
+    : thread_(thread_factory.createThread([this]() { work(); })) {}
+
+CacheEvictionThread::~CacheEvictionThread() {
```

Comment:
The pattern of calling ~CacheEvictionThread while CacheEvictionThread methods may be executing on another thread is subtly fragile. It's well-defined, but only because CacheEvictionThread is the most-derived class. Otherwise, any virtual methods running in another thread would cause undefined behavior.

Therefore, please either declare CacheSingleton as `final` (with a comment), or call `join` before ~CacheEvectionThread (also with a comment).

Reply:
I'm fairly sure you meant `CacheEvictionThread` as final - done.